src/Http/Http/src/Features/QueryFeature.cs

```diff
@@ -100,5 +98,227 @@ public IQueryCollection Query
+                        Uri.UnescapeDataString(name),
+                        Uri.UnescapeDataString(value));
+                }
+                else
+                {
+                    if (!querySegment.IsEmpty)
+                    {
+                        accumulator.Append(querySegment);
+                    }
+                }
+
+                if (delimiterIndex < 0)
+                {
+                    break;
+                }
+
+                query = query.Slice(delimiterIndex + 1);
+            }
+
+            return accumulator.HasValues
+                ? accumulator.GetResults()
+                : null;
+        }
+
+        internal struct KvpAccumulator
+        {
+            /// <summary>
+            /// This API supports infrastructure and is not intended to be used
+            /// directly from your code. This API may change or be removed in future releases.
+            /// </summary>
+            private AdaptiveCapacityDictionary<string, StringValues> _accumulator;
+            private AdaptiveCapacityDictionary<string, List<string>> _expandingAccumulator;
+
+            public void Append(ReadOnlySpan<char> key, ReadOnlySpan<char> value = default)
+                => Append(key.ToString(), value.IsEmpty ? string.Empty : value.ToString());
+
+            /// <summary>
+            /// This API supports infrastructure and is not intended to be used
+            /// directly from your code. This API may change or be removed in future releases.
+            /// </summary>
+            public void Append(string key, string value)
+            {
+                if (_accumulator is null)
+                {
+                    _accumulator = new AdaptiveCapacityDictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
+                }
+
+                if (_accumulator.TryGetValue(key, out var values))
+                {
+                    if (values.Count == 0)
+                    {
+                        // Marker entry for this key to indicate entry already in expanding list dictionary
+                        _expandingAccumulator[key].Add(value);
+                    }
+                    else if (values.Count == 1)
+                    {
+                        _accumulator[key] = StringValues.Concat(values, value);
+                    }
+                    else
+                    {
+                        // Add zero count entry and move to data to expanding list dictionary
+                        _accumulator[key] = default;
+
+                        if (_expandingAccumulator is null)
+                        {
+                            _expandingAccumulator = new AdaptiveCapacityDictionary<string, List<string>>(5, StringComparer.OrdinalIgnoreCase);
+                        }
+
+                        // Already 5 entries so use starting allocated as 10; then use List's expansion mechanism for more
+                        var list = new List<string>(10);
```

Comment:
? There should only be 3 entries at this point.

Reply:
2 already there + the new one = 3 ✔️ 

Is the capacity of 10 a good choice then?
How likely will it be to have queries with more than 3x the same key? Gut feeling tells me 6 is large enough.